fix: refactor `runConfigMapWatcher` to use Informers. Fixes #11657

juranir opened the pull request:
Fixes #11657 
Fixes #11463 

Supersedes / Closes #11799 

### Motivation

After some time running, when the connection to k8s API is broken (timeout for example), the watcher cannot handle the exception increases the CPU usage, and floods the log with the message `level=error msg="invalid config map object received in config watcher. Ignored processing"`. After reaching the pod limit, it'll be restarted. See #11657  for details.


### Modifications

I replaced the function with a new one using informers (many thanks @agilgur5 ).
I tried to use the code pattern used by other functions, but please let me know if something can be improved.
I also changed the namespace used to "managed-namespace" instead of "namespace" (This change will solve #11463 )

### Verification
Manual changes in the ConfigMap are being watched and applied;
An instance with this change is running for a while;
We also did some integration and performance tests;

![shell](https://github.com/argoproj/argo-workflows/assets/6254271/7c1fbabf-48d0-4148-b5b6-5b088af4eb73)

<!-- TODO: Say how you tested your changes. -->

juliev0 commented on workflow/controller/controller.go:
```diff
@@ -392,38 +395,34 @@ func (wfc *WorkflowController) initManagers(ctx context.Context) error {
+		FilterFunc: func(obj interface{}) bool {
+			cm, err := meta.Accessor(obj)
+			if err != nil {
+				return false
```
should we log something here?

juranir replied:
We can add a log here, even if it is as copy+past I also agree a log may be helpful.

juliev0 reviewed (commented):
just one small comment, but otherwise looks good!

agilgur5 commented:
> Fixes #11463

> I also changed the namespace used to "managed-namespace" instead of "namespace" (This change will solve #11463 )

I don't see how this fixes #11463? The RetryWatcher was already listening to `managedNamespace`.
#11463 also points out the opposite -- the Controller's ConfigMap is _not_ in the `managedNamespace`, it's in the Controller's own namespace. Using `managedNamespace` is the bug that #11463 points out.
Per my review comments in https://github.com/argoproj/argo-workflows/pull/11799#issuecomment-1714735706, there's also an issue with the semaphore notification that will have to be moved, as semaphores _are_ in the `managedNamespace`

agilgur5 commented on workflow/controller/controller.go:
```diff
@@ -392,38 +395,34 @@ func (wfc *WorkflowController) initManagers(ctx context.Context) error {
-			return wfc.kubeclientset.CoreV1().ConfigMaps(wfc.managedNamespace).Watch(ctx, metav1.ListOptions{})
-		},
+
+	informer := v1.NewConfigMapInformer(wfc.kubeclientset, wfc.GetManagedNamespace(), 20*time.Minute, cache.Indexers{
```
`wfc.GetManagedNamespace()` is better than `wfc.managedNamespace`, but still not quite correct, per my above comment. I'm fine if this PR doesn't fix the other bug, so either we can leave as is or we need to modify this to correctly fix the other bug

juliev0 replied:
I'd vote for changing it to `wfc.namespace` (thanks for noticing that @agilgur5)

juranir replied:
Great, I'll update it.

agilgur5 commented on workflow/controller/controller.go:
```diff
@@ -392,38 +395,34 @@ func (wfc *WorkflowController) initManagers(ctx context.Context) error {
+			return cm.GetName() == wfc.configController.GetName()
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(_, obj interface{}) {
```
technically speaking, an `AddFunc` and `DeleteFunc` would be good too, though I don't think the previous RetryWatcher handled those scenarios either. They are also more complex, so I think this PR is probably good as is

agilgur5 reviewed (changes requested):
There's a bit of an issue here with the semaphore configmap notifications. See in-line [below](https://github.com/argoproj/argo-workflows/pull/11855#discussion_r1341910749)

juranir commented on workflow/controller/controller.go:
```diff
@@ -109,6 +108,7 @@ type WorkflowController struct {
 	wftmplInformer        wfextvv1alpha1.WorkflowTemplateInformer
 	cwftmplInformer       wfextvv1alpha1.ClusterWorkflowTemplateInformer
 	podInformer           cache.SharedIndexInformer
+	configMapCtrlInformer cache.SharedIndexInformer
 	configMapInformer     cache.SharedIndexInformer
```
Great, I'll update it.

juranir commented on workflow/controller/controller.go:
```diff
@@ -392,38 +395,34 @@ func (wfc *WorkflowController) initManagers(ctx context.Context) error {
-			return
-		}
-	}
+				wfc.notifySemaphoreConfigUpdate(cm)
```
@agilgur5, I'm so sorry, but I'm not sure if I understood well your comment. Is it only about keeping `namespace` instead `managedNamespace`?

agilgur5 replied:
Not entirely.

`notifySemaphoreConfigUpdate` must run on the `managedNamespace`, yes, as all semaphore ConfigMaps are in the managed namespace.

Further, your `FilterFunc` limits the events to only the Controller's own ConfigMap, while `notifySemaphoreConfigUpdate` needs to run on _all_ ConfigMaps in the managed namespace.

The previous code before your modifications was kind of strangely coupled together (and had a bug for managed namespaces) as those are unrelated features, so I imagine you may have missed that one-line nuance

juliev0 replied:
@juranir there's some complexity here since [this](https://github.com/argoproj/argo-workflows/pull/11799) PR is also out at the same time. Both of you seem to have covered part of the equation, but then @agilgur5 has also given a lot of thought on how to best optimize and consolidate code [here](https://github.com/argoproj/argo-workflows/pull/11799#issuecomment-1742309369). It sounds like @agilgur5 is willing to take both of these PRs forward unless you would like to go through his comments in that issue and try to take it on yourself? (@agilgur5 please let me know if I'm mistaken)

agilgur5 replied:
yep that's correct. @juranir seemed interesting, so I'm very willing to assist him as well 🙂

juranir replied:
Thank you very much for all your help, @agilgur5  is very nice.

If I understand correctly, the idea is to refactor the current `newConfigMapInformer` function to handle both operations (plugins and the cm controller itself), so I think this [commit](https://github.com/argoproj/argo-workflows/pull/11855/commits/1a6110a7fae92ff3a6877f114f4c8564f4079164) might be on the way (I hope in the right one).

I also added a log as commented by @juliev0 

PS. One test failed, but I'm not sure about the reason or if it's related.

agilgur5 commented on workflow/controller/controller.go:
```diff
@@ -1053,42 +1017,62 @@ func (wfc *WorkflowController) newPodInformer(ctx context.Context) cache.SharedI
-	}, func(opts *metav1.ListOptions) {
-		opts.LabelSelector = common.LabelKeyConfigMapType
+
+	indexInformer := v1.NewConfigMapInformer(wfc.kubeclientset, wfc.GetManagedNamespace(), 20*time.Minute, cache.Indexers{
```
So I mentioned [before](https://github.com/argoproj/argo-workflows/pull/11855#discussion_r1342787539) that there are _two_ namespaces to watch, and both are used for different purposes. This is on the managed namespace, which is not necessarily the same as Argo's own namespace. The `workflow-controller-configmap` will _always_ be in Argo's own namespace, but semaphores etc may be in a separate namespace. As in, if you deploy Argo in one namespace but run Workflows in a _different_ namespace(s).

So we may very well need two separate Informers to handle both namespace configurations, unless you can listen to both namespaces in one.

juranir replied:
As far as I know, we can only watch 1 or all namespaces, so I think we have 2 possibilities (based on my limited knowledge):
#1. Create another function and duplicate the code;
#2. Create an array with all necessary namespaces and move the current code into this loop;

So far I would prefer #2, but I'm not sure if it may cause some problem

agilgur5 commented on workflow/controller/controller.go:
```diff
@@ -1053,42 +1017,62 @@ func (wfc *WorkflowController) newPodInformer(ctx context.Context) cache.SharedI
-				if err != nil {
-					return false
+
+	isPluginCM := true
```
the functions below can be called multiple times, so setting `isPluginCM` at the top-level here would be insufficient (and likely encounter race conditions as well). It would have to be set in each function individually.

juranir replied:
Do you think it would be better to convert this validation into a method and call it wherever we want?

agilgur5 replied:
That's one option, yea. `isPluginCM` as a function makes sense to me. Could do same for `isControllerCM` too.
They're both one-liners, but given the amount of times they're used here, helper functions would make it more readable and less error-prone.

agilgur5 commented on workflow/controller/controller.go:
```diff
@@ -1053,42 +1017,62 @@ func (wfc *WorkflowController) newPodInformer(ctx context.Context) cache.SharedI
-		opts.LabelSelector = common.LabelKeyConfigMapType
+
+	indexInformer := v1.NewConfigMapInformer(wfc.kubeclientset, wfc.GetManagedNamespace(), 20*time.Minute, cache.Indexers{
+		cache.NamespaceIndex: cache.MetaNamespaceIndexFunc,
```
I might be missing something here -- what was the purpose of changing the index?

agilgur5 commented on workflow/controller/controller.go:
```diff
@@ -1053,42 +1017,62 @@ func (wfc *WorkflowController) newPodInformer(ctx context.Context) cache.SharedI
+
+			if cm.GetName() == wfc.configController.GetName() {
+				isPluginCM = false
+			} else {
+				isPluginCM = true
+				log.WithField("executorPlugins", wfc.executorPlugins != nil).
+					Info("Plugins")
+			}
+
+			return (cm.GetName() == wfc.configController.GetName()) ||
+				(cm.GetLabels()[common.LabelKeyConfigMapType] == common.LabelValueTypeConfigMapExecutorPlugin)
```
so there is a _third_ type of ConfigMap that isn't handled here: the semaphore config update

juranir replied:
Sorry for asking, but I would like to clarify, `notifySemaphoreConfigUpdate` should be called for all CM caught by this informer or there is some exception (in the [current code](https://github.com/argoproj/argo-workflows/blob/master/workflow/controller/controller.go#L422) I think so)?

agilgur5 replied:
the current code is all CMs in the managed namespace. I don't think there's a specific label for them right now (but maybe there should be)

agilgur5 reviewed (changes requested):
Sorry for the delay on my end, I was sick for a whole week and then playing catch up.

There's two main issues with the below code, per the in-line comments:
1. There are potentially two different namespaces to watch, not just one
2. There are three different types of ConfigMaps to watch, not just two

tooptoop4 commented:
is this ok to be merged?

agilgur5 commented:
> is this ok to be merged?

No, my [previous review](https://github.com/argoproj/argo-workflows/pull/11855#pullrequestreview-1698131756) has yet to be addressed

agilgur5 commented:
Hey @juranir do you have any time to continue working on this next week? If not, I'd like to take over the work as #11657 has gotten to be one of the top issues (seemingly due to k8s 1.27 upgrades?).
I'll ofc credit you with the existing work and leave you as co-author etc 🙂

sstarcher commented:
@agilgur5 any updates on this it's becoming very problematic.

juranir committed:
fix: Refactor the func newConfigMapInformer to watch both main and mgmt ns. Fixes #11657

Signed-off-by: Juranir Santos <juranir.santos@gmail.com>

juranir committed:
Merge branch 'argoproj:main' into master

juranir commented:
The proposed suggestions by @agilgur5 and @juliev0  were implemented

juranir committed:
Merge branch 'main' into master

agilgur5 reviewed (changes requested):
Thanks for picking this up again @juranir! I had to get a solid block of time to re-familiarize myself with this code, but I reviewed everything now.

For other reviewers, note that turning on "Hide Whitespace" in the GH diff is very helpful (tbh, I am thinking of using a different reviewing UI entirely given how many complex PRs I review now and how many insufficiencies GH has).

This does look much better now. Summary of my in-line comments:
- Biggest thing -- this is [missing](https://github.com/argoproj/argo-workflows/pull/11855#discussion_r1554360641) the `runConfigMapWatcher` removal. That might've gotten lost in the rebase?
- I think [unconditionally](https://github.com/argoproj/argo-workflows/pull/11855#discussion_r1554393709) creating both Informers would simplify the code. If we have two Informers on the same namespace, so be it. Informer resource usage tends to be more in the cache of data and filtering, rather than the overhead of the Informer itself.
- there are [some](https://github.com/argoproj/argo-workflows/pull/11855#discussion_r1554387769) [missing](https://github.com/argoproj/argo-workflows/pull/11855#discussion_r1554386025) namespace checks
- the label selector on the configmap informer [needs moving](https://github.com/argoproj/argo-workflows/pull/11855#discussion_r1554396079)

agilgur5 commented:
While I have the context, and given the priority & impact of this, I'm just going to go ahead and make the changes requested above myself and push them to the PR

agilgur5 commented on workflow/controller/controller.go:
```diff
@@ -1270,43 +1285,62 @@ func (wfc *WorkflowController) newPodInformer(ctx context.Context) (cache.Shared
+	indexInformer := v1.NewFilteredConfigMapInformer(wfc.kubeclientset, ns, 20*time.Minute, cache.Indexers{
 		indexes.ConfigMapLabelsIndex: indexes.ConfigMapIndexFunc,
 	}, func(opts *metav1.ListOptions) {
 		opts.LabelSelector = common.LabelKeyConfigMapType
```
oh this is the indexer... this might be necessary then. and you already have the check in `isPluginCM()`, my bad

agilgur5 replied:
wait no it _is_ the `TweakListOptionsFunc` (have to look a bit closely 😅). so this should be removed

agilgur5 replied:
Some E2Es [failed](https://github.com/argoproj/argo-workflows/actions/runs/8578095350/job/23511562662?pr=11855) without this. They seemed to miss the label without this... would've thought it would've picked up all configmaps without this? 🤔 

I might be misunderstanding something here, I'm certainly not an Informer/`client-go` expert

agilgur5 replied:
Oh, nvm, totally different bug; the wrong informer was being used -- fixed in [this commit](https://github.com/argoproj/argo-workflows/pull/11855/commits/191b441ec3730782ff5963e7feccfd5df9597bbc)

Commit:
refactor out an applyPluginCM func

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
renames

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
move some functions around

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
add ns checks

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
comments

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
fix some more renames

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
also refactor out delete plugin for consistency

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
properly remove/replace runConfigMapWatcher

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
typings

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
woops not all those typings

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
use two configmap informers, one for managed ns, other for self ns

- they can overlap in one case, but this is fine
- we don't need a `FilteringResourceEventHandler` anymore:
  - for managed, we don't filter them, it's all configmaps in the ns
    - could potentially split two event handlers though?
  - for self, we have a field selector on the informer itself (which I believe is (much) more efficient)
- we no longer need the `isManagedNamespaceCM`, `isControllerCM` funcs
- nor `isManagedNamespaceDifferent`, `isCMInformerManagedSynced`

- the `WATCH_CONTROLLER_SEMAPHORE_CONFIGMAPS` flag is not very optimized, we could technically skip the entire self configmap informer with that now
  - but it may be removed as it was a workaround, so don't really want to add too much code for it
  - also if we add any other configmaps in the controller ns, we wouldn't have to change as much

- also properly pass `context` for the `UpdateConfig` call

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
remove extra spaces

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
field selector to string

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
fix test init

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
revert labelSelector removal? this seems to have broke some E2Es... hmmmm

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
Revert "revert labelSelector removal? this seems to have broke some E2Es... hmmmm"

This reverts commit 909b8e9b9fd4f9dff74939515c97ab735a1c92dc.

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
ah wrong informer

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

juranir commented:
> While I have the context, and given the priority & impact of this, I'm just going to go ahead and make the changes requested above myself and push them to the PR to make things faster (and reduce "telephone game" interpretations). In this review cycle I [now have permissions](https://github.com/argoproj/argoproj/issues/277) to do that.

No worries and many thanks for that.

@agilgur5  Please let me know if I can help with something.

Commit:
cm is obj just with a typecast, use the same arg

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
tiny refactor

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

agilgur5 commented on workflow/controller/controller.go:
```diff
@@ -1270,69 +1239,61 @@ func (wfc *WorkflowController) newPodInformer(ctx context.Context) (cache.Shared
+
+func (wfc *WorkflowController) newConfigMapInformerManaged() (cache.SharedIndexInformer, error) {
+	indexInformer := v1.NewConfigMapInformer(wfc.kubeclientset, wfc.GetManagedNamespace(), 20*time.Minute, cache.Indexers{
 		indexes.ConfigMapLabelsIndex: indexes.ConfigMapIndexFunc,
```
just noting here that this indexer [is used](https://github.com/argoproj/argo-workflows/blob/934bf179a703a365a6a5eddfa6edb788492eac79/workflow/controller/cache_gc.go#L28) for more than just plugins. that's why it previously only had a label selector on the label and did not specify the plugin type (and then why it filtered in the `FilterFunc` by plugin type instead -- it would populate the indexer but not the cache)

agilgur5 reviewed (commented):
I have substantially simplified this now (including refactoring out some common plugin handling functions) and fixed all the correctness issues. So this should be "good enough" now.

But:
1. There's a performance degradation, see in-line comment [below](https://github.com/argoproj/argo-workflows/pull/11855#discussion_r1556248023). **EDIT**: I have fixed that now with a third informer for semaphores
1. I'd like someone else to spot check this since I wrote a decent chunk of it. @juranir and @juliev0 can you both take a look?

Commit:
rename again -- all other informers don't specify 'managed' despite working on the managed ns. specify ctrl for controller instead

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
tiny optimization

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
rename miss fix

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

juranir commented on workflow/controller/controller.go:
```diff
@@ -1270,69 +1239,61 @@ func (wfc *WorkflowController) newPodInformer(ctx context.Context) (cache.Shared
+func (wfc *WorkflowController) newConfigMapInformerManaged() (cache.SharedIndexInformer, error) {
+	indexInformer := v1.NewConfigMapInformer(wfc.kubeclientset, wfc.GetManagedNamespace(), 20*time.Minute, cache.Indexers{
 		indexes.ConfigMapLabelsIndex: indexes.ConfigMapIndexFunc,
-	}, func(opts *metav1.ListOptions) {
-		opts.LabelSelector = common.LabelKeyConfigMapType
```
This option of getting a list of cm to process would be fantastic in terms of resource optimization. 

+1

agilgur5 replied:
Ok, I added the semaphore informer in [this commit](https://github.com/argoproj/argo-workflows/pull/11855/commits/f8e0872544602a96fae155f166d49709a32b28c2).

Now we're not changing the existing ConfigMap informer, which is nice, but tracking semaphores by name is a bit hacky 😕

agilgur5 commented on workflow/controller/controller.go:
```diff
@@ -1270,69 +1239,61 @@ func (wfc *WorkflowController) newPodInformer(ctx context.Context) (cache.Shared
+		UpdateFunc: func(_, obj interface{}) {
+			cm := obj.(*apiv1.ConfigMap)
+			log.Infof("Received Workflow Controller config map %s/%s update", cm.GetNamespace(), cm.GetName())
+			wfc.UpdateConfig(ctx)
```
there's an optimization that can be done here as well -- we can just pass the configmap directly. `UpdateConfig` does a `Get`, which is not necessary here.

that requires a bit of refactoring, including in the config controller, and is independent of this PR (the optimization probably could have been done before this PR as well), so I'm going to do that as a separate follow-up PR

Commit:
refactor out a third informer for semaphores, for better filtering

- this way the plugin etc informer is more efficient, only on configmaps with the label
  - and the diff is (much) smaller (once you ignore whitespace from the inverted `if`)

- semaphore informer only needs to watch certain configmaps -- get those names from the index
  - this is pretty hacky, I don't like this, but every way to track this seemed hacky 😕
    - this option had the least changes though
- semaphore informer only needs name and namespace -- remove the rest from its cache

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

agilgur5 commented on workflow/controller/indexes/workflow_index.go:
```diff
@@ -70,6 +71,16 @@ func WorkflowSemaphoreKeysIndexFunc() cache.IndexFunc {
 			return nil, nil
 		}
-		return wf.GetSemaphoreKeys(), nil
+		keys := wf.GetSemaphoreKeys()
+		// store all keys while indexing as a side effect
+		for _, key := range keys {
+			semaphoreKeyMap[key] = struct{}{}
```
This is pretty hacky, I don't like this, but every way to track this seemed hacky 😕
This option had the least changes though, which is why I ended up going with it

agilgur5 commented on workflow/controller/controller.go:
```diff
@@ -1274,69 +1251,155 @@ func (wfc *WorkflowController) newConfigMapInformer() (cache.SharedIndexInformer
+func (wfc *WorkflowController) newConfigMapSemaphoreInformer() (cache.SharedIndexInformer, error) {
+	indexInformer := v1.NewConfigMapInformer(wfc.kubeclientset, wfc.GetManagedNamespace(), 20*time.Minute, cache.Indexers{
+		cache.NamespaceIndex: cache.MetaNamespaceIndexFunc,
+	}).WithTransform(func(obj interface{}) (interface{}, error) {
+		cm, ok := obj.(*apiv1.ConfigMap)
+		if !ok {
+			return obj, nil
+		}
+
+		// only leave name and namespace, remove the rest as we don't use it
+		cm = cm.DeepCopy()
+		cm.ObjectMeta = metav1.ObjectMeta{
+			Name:      cm.Name,
+			Namespace: cm.Namespace,
 		}
+		cm.Data = map[string]string{}
+		cm.BinaryData = map[string][]byte{}
+		return cm, nil
+	})
```
here's [an external example](https://squiggly.dev/2023/06/custom-cache-for-kubernetes-controllers/) of using a `cache.TransformFunc`, for reference

Commit:
fix transform

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
woops i read the type wrong

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
we can remove some error handling as it will never happen

- the errors [only occur](https://github.com/kubernetes/client-go/blob/46588f2726fa3e25b1704d6418190f424f95a990/tools/cache/shared_informer.go#L580) when the informer was either already started or stopped
  - and we just created it, so we know that is not the case

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
Revert "we can remove some error handling as it will never happen"

This reverts commit b3e32746bd914c0db362df32761953d1bb381791.

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
fix unit tests

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
ignore the lint explicitly

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
remove some other unnecessary error handlers

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

agilgur5 commented on workflow/controller/controller.go:
```diff
@@ -1270,71 +1238,150 @@ func (wfc *WorkflowController) newPodInformer(ctx context.Context) (cache.Shared
-					delete(wfc.executorPlugins[namespace], name)
-					log.WithField("namespace", namespace).WithField("name", name).Info("Executor plugin removed")
-				},
+	//nolint:errcheck // the error only happens if the informer was stopped, and it hasn't even started (https://github.com/kubernetes/client-go/blob/46588f2726fa3e25b1704d6418190f424f95a990/tools/cache/shared_informer.go#L580)
```
see also https://github.com/argoproj/argo-workflows/pull/12847#discussion_r1556488443 -- will have a follow-up PR to remove the unnecessary error handling elsewhere.

a nice side-effect of this removal is that this will no longer merge conflict explicitly or implicitly during cherry-pick (as #12847 is breaking and may not be applied to 3.4.x/3.5.x)

agilgur5 reviewed (approved):
This LGTM to me now -- I've pretty much done every significant optimization that is specific and [isolated](https://github.com/argoproj/argo-workflows/pull/11855#discussion_r1556375760) to this change now
- `WATCH_CONTROLLER_SEMAPHORE_CONFIGMAPS` is not the most efficiently implemented (as I mentioned in the [initial commit](https://github.com/argoproj/argo-workflows/pull/11855/commits/43527e9c8073751cbd818b7bc48b7f14d90f5840)) as the Informers wouldn't need to be created at all in those cases (not just their event handlers), but given that's a workaround that could be removed (and env vars are explicitly documented as [unstable](https://argo-workflows.readthedocs.io/en/release-3.5/environment-variables/)), I didn't want to add too much code around it

There's also a comment I have [above](https://github.com/argoproj/argo-workflows/pull/11855#discussion_r1556460508) on the semaphore keys hackery in the indexer, but I don't think there's a better way to do that without a breaking change (e.g. requiring labels on semaphore configmaps)

I'm also going to build a test image and push to a test registry for folks to try and test out

agilgur5 commented:
> I'm also going to build a test image and push to a test registry for folks to try and test out

Pushed [`agilgur5/argo-workflow-controller:fix-configmap-informer`](https://hub.docker.com/layers/agilgur5/argo-workflow-controller/fix-configmap-informer/images/sha256-48c229a153ff18c9ab6fc94b4f62340e6155a6d68c25937c4ce733f73b444345?context=repo) for testing

juliev0 commented on workflow/controller/controller.go:
```diff
@@ -1270,71 +1238,150 @@ func (wfc *WorkflowController) newPodInformer(ctx context.Context) (cache.Shared
+	if _, ok := wfc.executorPlugins[cm.GetNamespace()]; !ok {
+		wfc.executorPlugins[cm.GetNamespace()] = map[string]*spec.Plugin{}
+	}
+	wfc.executorPlugins[cm.GetNamespace()][cm.GetName()] = p
```
Don't we need thread safety between these functions?

juliev0 replied:
(if so, I see it wasn't there before)

agilgur5 replied:
Yea this part hasn't changed in this PR, it just got refactored into a common func. If you turn on "Hide Whitespace" on the GH diff, this line actually appears as unchanged (I mentioned the setting [earlier](https://github.com/argoproj/argo-workflows/pull/11855#pullrequestreview-1984237287), although it's not as useful after my refactor)

Technically speaking, they probably should be made thread-safe in case an `apply` and `delete` happen simultaneously, but realistically that will basically never happen. Also, this is ran on a single goroutine, so there aren't multiple threads modifying it. Safety is good to have just in case though.

juliev0 replied:
It's run on a single goroutine? Oh, I figured the `ResourceEventHandlerFuncs` spawned new goroutines - they don't?

agilgur5 replied:
Oh sorry I missed the notification for this comment; just manually checked and saw it.

I thought it would run on the same goroutine as the Informer's `.Run`, but no I think you're correct, it does seem to spawn new goroutines. It was a little non-trivial to trace back, but this is [the `client-go` line](https://github.com/kubernetes/client-go/blob/aa7909e7d7c0661792ba21b9e882f3cd6ad0ce53/tools/cache/shared_informer.go#L802) that spawns goroutines (`AddEventHandler` calls [`processor.addListener`](https://github.com/kubernetes/client-go/blob/aa7909e7d7c0661792ba21b9e882f3cd6ad0ce53/tools/cache/shared_informer.go#L612) and `wait` `Group`'s `Start` [spawns it](https://github.com/kubernetes/apimachinery/blob/d794766488ac2892197a7cc8d0b4b46b0edbda80/pkg/util/wait/wait.go#L70)).

In any case though, as that thread-safety bug had existed beforehand, I don't plan to fix it in this PR. Adding locks would also conventionally entail creating a new, thread-safe data structure for the executor plugins which becomes a sizeable change in and of itself that detracts a bit from the purpose of this PR.
There's a couple other follow-ups from this PR stemming from incidental findings as well like https://github.com/argoproj/argo-workflows/pull/11855#discussion_r1556375760 and https://github.com/argoproj/argo-workflows/pull/11855#discussion_r1556460508

Have you reviewed the other parts of this PR?

juliev0 replied:
sounds good. Let me go through the rest of the PR tomorrow. :)

Commit:
fix wfc.Config.Namespace issue and add clarifying comment

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

juliev0 commented on workflow/controller/controller.go:
```diff
@@ -1270,71 +1238,150 @@ func (wfc *WorkflowController) newPodInformer(ctx context.Context) (cache.Shared
+			},
+			DeleteFunc: func(obj interface{}) {
+				cm := obj.(*apiv1.ConfigMap)
+				wfc.deletePluginCM(cm)
```
This is a lot cleaner looking :)

juliev0 commented on workflow/controller/controller.go:
```diff
@@ -1270,71 +1238,150 @@ func (wfc *WorkflowController) newPodInformer(ctx context.Context) (cache.Shared
+		opts.FieldSelector = fields.OneTermEqualSelector(metav1.ObjectNameField, wfc.configController.GetName()).String() // only the controller configmap
+	})
+
+	if !watchControllerSemaphoreConfigMaps {
```
I'm confused about this. Is it supposed to be here, or was it copy/pasted from below?

agilgur5 replied:
It is supposed to be there. If the flag is set, no event handlers are added to semaphore or controller informers.

They're together only because the flag was implemented prior to them being split (it was implemented as a workaround for the main bug this PR aims to fix)

juliev0 replied:
Oh jeez. Backward compatibility can be so annoying sometimes. :) I suppose some people do intend to not watch their Controller ConfigMap and are using that environment variable? In that case, can we make it more clear in the code through comments and/or naming that it represents both things?

agilgur5 replied:
> I suppose some people do intend to not watch their Controller ConfigMap and are using that environment variable?

Yes, https://github.com/argoproj/argo-helm/issues/2622 is one example, although I'm not committed to supporting that. I wanted to leave this flag in just-in-case there are remaining issues. The hot reload has also been buggy before, e.g. #11177, #11542, #11541, so it is potentially useful as a workaround if that reoccurs.

> In that case, can we make it more clear in the code through comments and possibly naming that it represents both things?

Sure I can add a comment. I made sure it was described in the docs in #12622 and #12400, but code comments are missing.

The naming follows the env variable exactly as the codebase conventionally does, so I wouldn't change that though.

juliev0 replied:
okay, sounds good

agilgur5 replied:
[added](https://github.com/argoproj/argo-workflows/pull/11855/commits/05ee735a985473e4f15b1fc642f6ff96571aa7c5) comment

Also fixed merge conflicts (it was with my own PR that stemmed from a finding in this PR: #12917. just had to choose this PR's version for each conflict)

juliev0 commented:
Unfortunately, it looks like there are some file conflicts now.

Another question: which of the different configmaps have been effectively tested (either manually, unit testing, or e2e)?

Commit:
add code comment about env var

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

Commit:
Merge branch 'main' into fix-configmap-informer

Signed-off-by: Anton Gilgur <agilgur5@gmail.com>

juliev0 commented:
I'm going through old PRs now, looking for stale ones to add "stale" label to. What's the status of this one @agilgur5 @juranir ?

agilgur5 commented:
It's still very necessary, although the upstream issue may have been resolved in k8s 1.30 if our hypotheses are correct, but even if that is accurate it may break again as it was due to a breaking bugfix in etcd that is now feature flagged due to the unexpected breakage. This change is still good to have and fixes other issues either way as well.

But I never got around to writing the tests, and, as far as I could tell, there were no tests for this previously, which would explain why it wasn't caught during k8s upgrades.

Joibel commented:
Replaced by #16408
